compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -915,35 +916,35 @@ recExpFields (EStructProj f e) = do
     Nothing -> return Nothing
 recExpFields _ = return Nothing
 
-instance SubstTm BoolExpr where
-  substituteTm s (BExpr e) = BExpr (substituteTm s e)
-  substituteTm s (BAnd e1 e2) = BAnd (substituteTm s e1) (substituteTm s e2)
-  substituteTm s (BNot e) = BNot (substituteTm s e)
-  substituteTm s (BEq e1 e2) = BEq (substituteTm s e1) (substituteTm s e2)
-  substituteTm s (BGt e1 e2) = BGt (substituteTm s e1) (substituteTm s e2)
-  substituteTm s (BGtE e1 e2) = BGtE (substituteTm s e1) (substituteTm s e2)
-  substituteTm s (BLt e1 e2) = BLt (substituteTm s e1) (substituteTm s e2)
-  substituteTm s (BLtE e1 e2) = BLtE (substituteTm s e1) (substituteTm s e2)
-
-instance SubstTm a => SubstTm (Cond a) where
-  substituteTm s (Determined x) = Determined $ substituteTm s x
-  substituteTm s (Conditional e x y) =
-    Conditional (substituteTm s e) (map (substituteTm s) x) (map (substituteTm s) y)
-
-instance SubstTm a => SubstTm (Rec a) where
-  substituteTm s = \case
-    Simple x -> Simple (substituteTm s x)
-    Rec xs -> Rec (map (substituteTm s) xs)
-    MutRec xs -> MutRec (map (second (substituteTm s)) xs)
-
-instance IsPhase ph => SubstTm (Upd ph) where
-  substituteTm s = mapBaseUpd (baseUpd . substituteTm s)
-
-instance SubstTm BaseUpd where
-  substituteTm s UpdCreate{..} = UpdCreate _creTemp
-    (map (second (substituteTm s)) _creField)
-  substituteTm s UpdArchive{..} = UpdArchive _arcTemp
-    (map (second (substituteTm s)) _arcField)
+applySubstInBoolExpr :: Subst -> BoolExpr -> BoolExpr
+applySubstInBoolExpr subst = \case
+    BExpr e -> BExpr (applySubstInExpr subst e)
+    BAnd e1 e2 -> BAnd (applySubstInBoolExpr subst e1) (applySubstInBoolExpr subst e2)
+    BNot e -> BNot (applySubstInBoolExpr subst e)
+    BEq e1 e2 -> BEq (applySubstInExpr subst e1) (applySubstInExpr subst e2)
+    BGt e1 e2 -> BGt (applySubstInExpr subst e1) (applySubstInExpr subst e2)
+    BGtE e1 e2 -> BGtE (applySubstInExpr subst e1) (applySubstInExpr subst e2)
+    BLt e1 e2 -> BLt (applySubstInExpr subst e1) (applySubstInExpr subst e2)
+    BLtE e1 e2 -> BLtE (applySubstInExpr subst e1) (applySubstInExpr subst e2)
+
+applySubstInCond :: (Subst -> a -> a) -> Subst -> Cond a -> Cond a
```

Comment:
This is going towards a direction where a typeclass might come in handy but given that this is internal to this module it didn’t seem worth it.

Reply:
I think this combinator style is perfectly fine.